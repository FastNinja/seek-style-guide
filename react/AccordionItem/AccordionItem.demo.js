import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { AccordionItem, Text, PageBlock, Section } from 'seek-style-guide/react';

class AccordionContainer extends Component {
  static propTypes = {
    component: PropTypes.func.isRequired,
    componentProps: PropTypes.object.isRequired
  };

  render() {
    const { component: DemoComponent, componentProps } = this.props;
    return (
      <PageBlock>
        <Section>
          <DemoComponent
            {...componentProps}
          />
        </Section>
      </PageBlock>
    );
  }
}

export default {
  route: '/accordionitem',
  title: 'Accordion Item',
  category: 'Layout',
  component: AccordionItem,
  container: AccordionContainer,
  block: true,
  initialProps: {
    title: <Text baseline={false} raw>Living Style Guide</Text>,
    id: 'accordion_1',
    children: <Text>This is some text inside of an Accordion Item</Text>
  },
  options: [
    {
      label: 'Type',
      type: 'checkbox',
      states: [
        {
          label: 'Expanded',
          transformProps: props => ({
            ...props,
            expanded: true
          })
        }
      ]
    },
    {
      label: 'Type',
      type: 'checkbox',
      states: [
        {
          label: 'Custom onClick',
          transformProps: props => ({
            ...props,
            onClick: expanded => console.log(expanded)
          })
        }
      ]
    }
  ]
};
