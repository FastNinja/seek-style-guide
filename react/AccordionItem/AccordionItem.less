@import (reference) "~seek-style-guide/theme";

.toggle {
  position: absolute;
  opacity: 0;
}

.toggleLink {
  display: flex;
  justify-content: space-between;
  align-items: center;
  color: @sk-link;
  cursor: pointer;
  min-height: (@grid-row-height * 4);
}

.toggleContainer {
  display: none;
}

.chevron {
  display: inline-block;
  margin-left: 9px;
  transition: transform 0.3s ease;
  transform: translateY(1px);
  transform-origin: 50% 50%;
}

.chevronSvg {
  width: 12px;
  height: 12px;
}

.toggle:checked ~ .toggleContainer {
  display: block;
}

.toggle:checked ~ .toggleLabel .chevron {
  transform: rotate(180deg);
}
