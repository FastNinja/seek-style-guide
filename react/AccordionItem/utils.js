import { CLOSED_HEIGHT, DURATION } from './constants';

function onAnimationEnd(fn) {
  return setTimeout(fn, DURATION);
}

export function toggleContent({
  el,
  setCurrentHeight,
  timeoutHandle,
  setTimeoutHandle,
  isOpen,
  setIsOpen,
  setCssVisibility,
  setCssOverflow,
  setCssOpacity
}) {
  const contentHeight = el.offsetHeight;
  const isOpening = !isOpen;
  console.log('isOpening', isOpening);
  clearTimeout(timeoutHandle);

  if (isOpening) {
    setCssVisibility('visible');
    setCurrentHeight(`${contentHeight}px`);
    setIsOpen(true);
    setCssOpacity(1);

    setTimeoutHandle(
      onAnimationEnd(() => {
        // We toggle overflow on and off to avoid cropping the focus state
        // on form fields
        setCssOverflow('visible');
      })
    );
  } else {
    console.log('contentHeight', contentHeight);
    setCurrentHeight(CLOSED_HEIGHT);
    setCssOverflow('hidden');
    setIsOpen(false);
    setCssOpacity(0);

    setTimeoutHandle(
      onAnimationEnd(() => {
        setCssVisibility('hidden');
      })
    );
  }
}
