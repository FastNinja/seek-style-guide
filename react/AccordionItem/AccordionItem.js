import React, { Fragment, useRef, useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { ChevronIcon, TextLink, Text } from 'seek-style-guide/react';
import classnames from 'classnames';
import styles from './AccordionItem.less';
import { toggleContent } from './utils';
import { CLOSED_HEIGHT } from './constants';

function AccordionItem({
  className,
  title,
  children,
  open,
  isOpen: externalIsOpen,
  setIsOpen: externalSetIsOpen,
  onClick,
  onOpen,
  onClose,
  ...restProps
}) {
  const initialHeight = open ? 'auto' : CLOSED_HEIGHT;
  const initialVisibility = open ? 'visible' : 'hidden';
  const initialOverflow = initialVisibility;
  const contentEl = useRef(null);
  const [currentHeight, setCurrentHeight] = useState(initialHeight);
  const [cssVisibility, setCssVisibility] = useState(initialVisibility);
  const [cssOverflow, setCssOverflow] = useState(initialOverflow);
  const [timeoutHandle, setTimeoutHandle] = useState(null);
  const [isOpen, setIsOpen] = useState(open);
  const useInternalState = externalIsOpen === undefined;
  const finalIsOpen = useInternalState ? isOpen : !externalIsOpen;

  useEffect(() => {
    if (!useInternalState) {
      toggleContent({
        el: contentEl.current,
        setCurrentHeight,
        timeoutHandle,
        setTimeoutHandle,
        isOpen: finalIsOpen,
        setIsOpen: () => {},
        setCssVisibility,
        setCssOverflow
      });
    }
  }, [finalIsOpen]);

  const buttonClasses = classnames(className, styles.title);
  const isAnimating = currentHeight === CLOSED_HEIGHT && isOpen;
  const expanderClasses = classnames(styles.expander, {
    [styles.expanderOpen]: isAnimating
  });
  const chevronDirection = isOpen ? 'up' : 'down';

  return (
    <Fragment>
      <button
        type="button"
        className={buttonClasses}
        onClick={() => {
          if (useInternalState) {
            toggleContent({
              el: contentEl.current,
              setCurrentHeight,
              timeoutHandle,
              setTimeoutHandle,
              isOpen: finalIsOpen,
              setIsOpen,
              setCssVisibility,
              setCssOverflow
            });
          }

          if (onClick) {
            onClick();
          }

          if (isOpen && onClose) {
            onClose();
          } else if (!isOpen && onOpen) {
            onOpen();
          }
        }}
        {...restProps}
      >
        {typeof title === 'string' ? (
          <Text raw baseline={false} className={styles.titleText}>
            <TextLink>{title}</TextLink>
          </Text>
        ) : (
          <span>{title}</span>
        )}
        <ChevronIcon
          direction={chevronDirection}
          className={styles.chevron}
          svgClassName={styles.chevronSvg}
        />
      </button>
      <div
        className={expanderClasses}
        style={{ height: currentHeight, overflow: cssOverflow }}
      >
        <div
          className={styles.content}
          ref={contentEl}
          style={{ visibility: cssVisibility }}
        >
          {children}
        </div>
      </div>
    </Fragment>
  );
}

function usePrevious(value) {
  const ref = useRef();
  useEffect(() => {
    ref.current = value;
  });
  return ref.current;
}

AccordionItem.propTypes = {
  title: PropTypes.oneOfType([PropTypes.element, PropTypes.string]).isRequired,
  onOpen: PropTypes.func,
  onClose: PropTypes.func,
  children: PropTypes.element.isRequired,
  className: PropTypes.string,
  open: PropTypes.bool
};

AccordionItem.defaultProps = {
  className: '',
  open: false
};

export default AccordionItem;
