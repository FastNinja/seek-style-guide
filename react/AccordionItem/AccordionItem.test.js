import React from 'react';
import { shallow } from 'enzyme';
import AccordionItem from './AccordionItem';

const defaultProps = {
  title: () => <span>Living Style Guide</span>,
  children: <span>This is the content inside an Accordion Item</span>,
  id: 'abc123'
}
const renderAccordion = (props = defaultProps) => shallow(<AccordionItem {...props}>Hello world</AccordionItem>);

test('should render AccordionItem', () => {
  const accordion = renderAccordion();
  expect(accordion).toMatchSnapshot();
});

test('should render with expanded prop', () => {
  const accordion = renderAccordion({ ...defaultProps, expanded: true });
  expect(accordion).toMatchSnapshot();
});

test('should render with className', () => {
  const accordion = renderAccordion({ ...defaultProps, className: 'foobar' });
  expect(accordion).toMatchSnapshot();
});

test('should render expanded when prop changes', () => {
  jest.spyOn(AccordionItem.prototype, 'componentWillReceiveProps');
  const accordion = renderAccordion();
  accordion.setProps({ expanded: true });
  expect(AccordionItem.prototype.componentWillReceiveProps.mock.calls.length).toBe(1);
  expect(accordion.find({ type: 'checkbox' }).props().checked).toBe(true);
  expect(accordion).toMatchSnapshot();
});
