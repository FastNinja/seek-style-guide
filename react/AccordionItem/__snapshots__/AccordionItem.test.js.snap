// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AccordionItem 1`] = `
<div
  className=""
>
  <input
    checked={false}
    className="AccordionItem__toggle"
    id="abc123"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    className="AccordionItem__toggleLink AccordionItem__toggleLabel"
    htmlFor="abc123"
  >
    <Component />
    <ChevronIcon
      className="AccordionItem__chevron"
      direction="down"
      svgClassName="AccordionItem__chevronSvg"
    />
  </label>
  <div
    className="AccordionItem__toggleContainer"
  >
    Hello world
  </div>
</div>
`;

exports[`should render expanded when prop changes 1`] = `
<div
  className=""
>
  <input
    checked={true}
    className="AccordionItem__toggle"
    id="abc123"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    className="AccordionItem__toggleLink AccordionItem__toggleLabel"
    htmlFor="abc123"
  >
    <Component />
    <ChevronIcon
      className="AccordionItem__chevron"
      direction="down"
      svgClassName="AccordionItem__chevronSvg"
    />
  </label>
  <div
    className="AccordionItem__toggleContainer"
  >
    Hello world
  </div>
</div>
`;

exports[`should render with className 1`] = `
<div
  className="foobar"
>
  <input
    checked={false}
    className="AccordionItem__toggle"
    id="abc123"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    className="AccordionItem__toggleLink AccordionItem__toggleLabel"
    htmlFor="abc123"
  >
    <Component />
    <ChevronIcon
      className="AccordionItem__chevron"
      direction="down"
      svgClassName="AccordionItem__chevronSvg"
    />
  </label>
  <div
    className="AccordionItem__toggleContainer"
  >
    Hello world
  </div>
</div>
`;

exports[`should render with expanded prop 1`] = `
<div
  className=""
>
  <input
    checked={true}
    className="AccordionItem__toggle"
    id="abc123"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    className="AccordionItem__toggleLink AccordionItem__toggleLabel"
    htmlFor="abc123"
  >
    <Component />
    <ChevronIcon
      className="AccordionItem__chevron"
      direction="down"
      svgClassName="AccordionItem__chevronSvg"
    />
  </label>
  <div
    className="AccordionItem__toggleContainer"
  >
    Hello world
  </div>
</div>
`;
